File: ts/src/parse.ts
import * as YAML from "yaml";
import { parse as parseTomlText } from "smol-toml";

import { EntryconfError } from "./errors.ts";
import { emptyObject, setKey, type Value } from "./tree.ts";

export type Format = "json" | "yaml" | "toml";

const EXTENSIONS: Record<string, Format> = {
  ".json": "json",
  ".yaml": "yaml",
  ".yml": "yaml",
  ".toml": "toml",
};

/**
 * Parser selected by file extension (SPEC §5); null for unsupported ones.
 * The match is case-sensitive: `.JSON` is "any other extension".
 */
export function formatForExtension(ext: string): Format | null {
  return Object.hasOwn(EXTENSIONS, ext) ? EXTENSIONS[ext] : null;
}

export function stripBom(text: string): string {
  return text.charCodeAt(0) === 0xfeff ? text.slice(1) : text;
}

/**
 * Decode file bytes as UTF-8, strictly. `readFileSync(path, "utf8")` silently
 * substitutes U+FFFD for invalid bytes; SPEC §2 makes content that is not
 * valid UTF-8 an `E_PARSE` anywhere it appears.
 */
export function decodeUtf8(bytes: Uint8Array, path: string): string {
  try {
    return new TextDecoder("utf-8", { fatal: true }).decode(bytes);
  } catch {
    throw parseError(path, "content is not valid UTF-8");
  }
}

function parseError(path: string, detail: string): EntryconfError {
  return new EntryconfError("E_PARSE", `${path}: ${detail}`);
}

export function parseDocument(
  text: string,
  format: Format,
  path: string,
): Value {
  const source = stripBom(text);
  switch (format) {
    case "json":
      return normalize(parseJson(source, path), path, false);
    case "yaml":
      // Only YAML has aliases, so only YAML carries the expansion budget
      // (SPEC §2); a legitimately huge JSON or TOML document is not bounded.
      return normalize(parseYaml(source, path), path, true);
    case "toml":
      return normalize(parseToml(source, path), path, false);
  }
}

// --- JSON ------------------------------------------------------------------

function parseJson(text: string, path: string): unknown {
  let value: unknown;
  try {
    value = JSON.parse(text);
  } catch (err) {
    throw parseError(path, (err as Error).message);
  }
  // JSON.parse is last-wins on duplicate keys; SPEC §2 makes them E_PARSE, so
  // scan the (already validated) token stream for repeated keys per object.
  assertNoDuplicateJsonKeys(text, path);
  return value;
}

function assertNoDuplicateJsonKeys(text: string, path: string): void {
  // `text` has already been accepted by JSON.parse, so the scan below can rely
  // on the document being well-formed: a string token followed by `:` while an
  // object is open is a key, and nothing else is.
  const stack: (Set<string> | null)[] = [];
  let i = 0;
  while (i < text.length) {
    const c = text[i];
    if (c === "{") {
      stack.push(new Set<string>());
      i++;
    } else if (c === "[") {
      stack.push(null);
      i++;
    } else if (c === "}" || c === "]") {
      stack.pop();
      i++;
    } else if (c === '"') {
      const end = scanJsonString(text, i);
      const raw = text.slice(i, end);
      i = end;
      let j = i;
      while (j < text.length && isJsonWhitespace(text[j])) j++;
      const keys = stack.length > 0 ? stack[stack.length - 1] : null;
      if (text[j] === ":" && keys) {
        const key = JSON.parse(raw) as string;
        if (keys.has(key)) {
          throw parseError(path, `duplicate key ${JSON.stringify(key)}`);
        }
        keys.add(key);
      }
    } else {
      i++;
    }
  }
}

function isJsonWhitespace(c: string): boolean {
  return c === " " || c === "\t" || c === "\n" || c === "\r";
}

/** Index just past the closing quote of the string token starting at `start`. */
function scanJsonString(text: string, start: number): number {
  let i = start + 1;
  while (i < text.length) {
    const c = text[i];
    if (c === "\\") {
      i += 2;
      continue;
    }
    if (c === '"') return i + 1;
    i++;
  }
  return text.length;
}

// --- YAML ------------------------------------------------------------------

function parseYaml(text: string, path: string): unknown {
  const docs = YAML.parseAllDocuments(text, {
    logLevel: "silent",
    schema: "core",
    version: "1.2",
    // Duplicate keys are E_PARSE (SPEC §2), but the `yaml` package's own
    // `uniqueKeys` check compares each new key against every key already in
    // the mapping — quadratic, and ruinous on a large alias-free mapping
    // (hundreds of seconds for a few hundred thousand entries). Detect them
    // linearly in `assertYamlKeys` instead, which already visits every Pair.
    uniqueKeys: false,
    merge: false,
    intAsBigInt: false,
  });
  if (docs.length > 1) {
    throw parseError(path, "multi-document YAML streams are not supported");
  }
  if (docs.length === 0) return null;
  const doc = docs[0];
  if (doc.errors.length > 0) {
    const first = doc.errors[0];
    throw parseError(path, first.message);
  }
  // Unresolvable (custom) tags are warnings in the `yaml` package; SPEC §2
  // makes them E_PARSE.
  for (const warning of doc.warnings) {
    if (warning.code === "TAG_RESOLVE_FAILED") {
      throw parseError(path, warning.message);
    }
  }
  // Runs on the AST, before `toJS` expands aliases and before the node budget
  // is charged, so a document that is both duplicate-keyed and oversized is
  // rejected on the key — a dup-key bomb cannot slip past the check by being
  // expensive to expand.
  assertYamlKeys(doc, path);
  try {
    // `maxAliasCount` is a *toJS* option, and its default (100) counts alias
    // references rather than expanded nodes: it rejects honest heavy reuse
    // (case 58 makes 109 references expanding to well under a thousand nodes)
    // while saying nothing about the actual expanded size. Disable it and let
    // `normalize` enforce the real budget SPEC §2 specifies. With aliases
    // unresolved into copies, `toJS` returns a cheap shared graph — repeated
    // aliases to one anchor yield the *same* object — so the expansion (and
    // the blowup an alias bomb is after) happens in `normalize`, which is
    // where the node count has to be charged.
    return doc.toJS({ maxAliasCount: -1 });
  } catch (err) {
    throw parseError(path, (err as Error).message);
  }
}

/**
 * Enforce the two key rules of SPEC §2 in one AST walk: mapping keys MUST be
 * strings, and a duplicate key within one document is `E_PARSE`.
 *
 * Both checks run on the parsed AST, where every anchor appears exactly once
 * and an alias is a single node, so the walk costs the size of the *source*
 * document rather than the size of its expansion — and it is linear in that
 * size: each mapping's keys go into a `Set` of strings, one lookup per entry,
 * in place of the `yaml` package's pairwise `uniqueKeys` comparison.
 */
function assertYamlKeys(doc: YAML.Document.Parsed, path: string): void {
  YAML.visit(doc, {
    Map(_key, node) {
      const seen = new Set<string>();
      for (const pair of node.items) {
        const k = pair.key;
        // Non-string keys are rejected by the `Pair` visitor below, which this
        // pre-order walk reaches right after this mapping; skip them here so
        // the two checks cannot disagree about what a key's text is.
        if (!YAML.isScalar(k) || typeof k.value !== "string") continue;
        if (seen.has(k.value)) {
          throw parseError(path, `duplicate key ${JSON.stringify(k.value)}`);
        }
        seen.add(k.value);
      }
    },
    Pair(_key, pair) {
      const k = pair.key;
      if (!YAML.isScalar(k) || typeof k.value !== "string") {
        throw parseError(
          path,
          "mapping keys must be strings for a JSON-equivalent tree",
        );
      }
    },
  });
}

// --- TOML ------------------------------------------------------------------

function parseToml(text: string, path: string): unknown {
  assertNoTomlOffsetTime(text, path);
  try {
    return parseTomlText(text);
  } catch (err) {
    throw parseError(path, (err as Error).message);
  }
}

/**
 * TOML 1.0 has no offset-time type — an offset is only valid after a full
 * date-time — but smol-toml accepts `07:32:00+05:00`, silently shifts the
 * clock, and returns a local-time `TomlDate` that keeps no trace of the
 * offset (SPEC §2 makes such input `E_PARSE`). The `TomlDate` is
 * indistinguishable from a genuine local time, so the check must run on the
 * source text: blank out strings and comments, then flag any offset-time
 * token that is not preceded by a date and separator.
 */
function assertNoTomlOffsetTime(text: string, path: string): void {
  const scannable = blankTomlStringsAndComments(text);
  const offsetTime = /\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:[Zz]|[+-]\d{2}:\d{2})/g;
  for (const match of scannable.matchAll(offsetTime)) {
    const start = match.index;
    const inDateTime =
      start >= 2 &&
      /[Tt ]/.test(scannable[start - 1]) &&
      /\d/.test(scannable[start - 2]);
    if (!inDateTime) {
      throw parseError(
        path,
        `${match[0]} is not a TOML value (a time offset requires a full date-time)`,
      );
    }
  }
}

/**
 * Replace the contents of TOML strings and comments with spaces (newlines
 * kept), so token-level scans cannot be fooled by string values. Handles
 * basic/literal strings, their multiline forms (including up to two closing
 * quotes owned by the content), and `\` escapes in basic strings. On
 * malformed input this errs toward blanking too much — smol-toml still
 * governs the actual parse, so the worst case is a missed pre-check, never a
 * false E_PARSE.
 */
function blankTomlStringsAndComments(text: string): string {
  let out = "";
  let i = 0;
  while (i < text.length) {
    const c = text[i];
    if (c === "#") {
      while (i < text.length && text[i] !== "\n") {
        out += " ";
        i++;
      }
      continue;
    }
    if (c === '"' || c === "'") {
      const quote = c;
      const triple = text.startsWith(quote.repeat(3), i);
      const delim = triple ? quote.repeat(3) : quote;
      out += " ".repeat(delim.length);
      i += delim.length;
      while (i < text.length) {
        if (quote === '"' && text[i] === "\\") {
          out += "  ";
          i += 2;
          continue;
        }
        if (text.startsWith(delim, i)) {
          if (triple) {
            // A multiline delimiter may be preceded by 1-2 quotes that belong
            // to the content ("""" ends a string whose content ends in `"`).
            let extra = 0;
            while (extra < 2 && text[i + delim.length + extra] === quote) {
              extra++;
            }
            out += " ".repeat(extra);
            i += extra;
          }
          out += " ".repeat(delim.length);
          i += delim.length;
          break;
        }
        out += text[i] === "\n" ? "\n" : " ";
        i++;
      }
      continue;
    }
    out += c;
    i++;
  }
  return out;
}

/**
 * smol-toml returns a `TomlDate` (a `Date` subclass) whose `toISOString()`
 * preserves the authored shape — local date, local time, local date-time, or
 * offset date-time — but keeps the source's own spelling of the separator and
 * the offset, and always renders milliseconds. SPEC §2 pins one rendering:
 * uppercase `T` separator, a UTC offset (`Z`, `z`, or `+00:00`) written as
 * `Z`, any other offset in its authored numeric form, and fractional seconds
 * with trailing zeros dropped (the `.` going with them at zero).
 */
function tomlDateToString(date: Date): string {
  let text = date.toISOString();

  // TOML permits "t"/" " as the date/time separator; the spec pins "T".
  text = text.replace(/^(\d{4}-\d{2}-\d{2})[tT ](?=\d{2}:)/, "$1T");

  // Split the offset off so it cannot be confused with a time fragment.
  let offset = "";
  const found = /(?:[zZ]|[+-]\d{2}:\d{2})$/.exec(text);
  if (found) {
    const authored = found[0];
    text = text.slice(0, text.length - authored.length);
    offset =
      authored === "z" || authored === "Z" || authored === "+00:00"
        ? "Z"
        : authored;
  }

  // Fractional seconds: drop trailing zeros, and the "." at zero.
  text = text.replace(
    /(\d{2}:\d{2}:\d{2})\.(\d+)$/,
    (_all, time: string, fraction: string) => {
      const trimmed = fraction.replace(/0+$/, "");
      return trimmed === "" ? time : `${time}.${trimmed}`;
    },
  );

  return text + offset;
}

// --- normalization ---------------------------------------------------------

/**
 * SPEC §2: a YAML document whose fully expanded tree would exceed this many
 * nodes is `E_PARSE`. Each scalar value, sequence element, and mapping entry
 * counts as one.
 */
export const MAX_EXPANDED_NODES = 1_000_000;

/** State threaded through the normalization walk. */
interface NormalizeContext {
  readonly path: string;
  /**
   * The chain of collections currently being walked, so an alias pointing at
   * one of its own ancestors — a cyclic structure rather than a plain value —
   * is caught. Entries are removed on the way back out, so a node reachable
   * many times over (the normal alias case) is not mistaken for a cycle.
   */
  readonly ancestors: Set<object>;
  /** Nodes still affordable, or null when no budget applies (JSON/TOML). */
  budget: number | null;
}

/**
 * Convert a parser's output into the JSON-equivalent data model, rejecting
 * anything that is not representable (SPEC §2).
 *
 * When `bounded` (YAML), the walk also charges the alias-expansion budget.
 * The walk *is* the expansion: `toJS` hands back a graph in which every alias
 * to one anchor is the same object, and rebuilding it here into a plain tree
 * duplicates those shared structures. So counting nodes as they are produced
 * measures exactly what SPEC §2 bounds — the size of the expanded tree — and
 * an alias bomb is stopped after a million nodes instead of materializing all
 * 48 million of them.
 */
function normalize(value: unknown, path: string, bounded: boolean): Value {
  return normalizeValue(value, {
    path,
    ancestors: new Set(),
    budget: bounded ? MAX_EXPANDED_NODES : null,
  });
}

/** Charge one node against the budget, failing once it is spent. */
function charge(ctx: NormalizeContext): void {
  if (ctx.budget === null) return;
  if (ctx.budget === 0) {
    throw parseError(
      ctx.path,
      `alias expansion exceeds the ${MAX_EXPANDED_NODES}-node budget`,
    );
  }
  ctx.budget -= 1;
}

function normalizeValue(value: unknown, ctx: NormalizeContext): Value {
  const path = ctx.path;
  if (value === null || value === undefined) {
    charge(ctx);
    return null;
  }
  switch (typeof value) {
    case "boolean":
    case "string":
      charge(ctx);
      return value;
    case "number":
      if (!Number.isFinite(value)) {
        throw parseError(
          path,
          `${String(value)} is not representable in the JSON-equivalent data model`,
        );
      }
      charge(ctx);
      return value;
    case "bigint":
      throw parseError(path, "integer is out of range for a JSON number");
    case "object":
      break;
    default:
      throw parseError(path, `unsupported value of type ${typeof value}`);
  }
  if (value instanceof Date) {
    charge(ctx);
    return tomlDateToString(value);
  }

  const node = value as object;
  if (ctx.ancestors.has(node)) {
    throw parseError(path, "recursive alias does not resolve to a plain value");
  }
  ctx.ancestors.add(node);
  try {
    if (Array.isArray(value)) {
      const items: Value[] = [];
      for (const item of value) {
        charge(ctx); // the sequence element
        items.push(normalizeValue(item, ctx));
      }
      return items;
    }
    const out = emptyObject();
    for (const [key, item] of Object.entries(node)) {
      charge(ctx); // the mapping entry
      setKey(out, key, normalizeValue(item, ctx));
    }
    return out;
  } finally {
    ctx.ancestors.delete(node);
  }
}

File: go/parse.go
package entryconf

import (
	"bytes"
	"encoding/json"
	"errors"
	"io"
	"math"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"
	"unicode/utf8"

	toml "github.com/pelletier/go-toml/v2"
	yaml "gopkg.in/yaml.v3"
)

// parserFor maps a file extension to a parser (SPEC §5). Extensions are
// matched case-sensitively in their lowercase spelling, so ".JSON" is not a
// config document.
func parserFor(path string) (func(path string, data []byte) (any, error), bool) {
	switch filepath.Ext(path) {
	case ".json":
		return parseJSON, true
	case ".yaml", ".yml":
		return parseYAML, true
	case ".toml":
		return parseTOML, true
	}
	return nil, false
}

// parseDocument reads and parses a document. A missing or unreadable file is
// reported by the caller (which knows whether it is an entrypoint or an
// include); an unparseable one — including one whose bytes are not valid
// UTF-8 (SPEC §2) — is always E_PARSE.
func parseDocument(path string) (any, error) {
	parse, ok := parserFor(path)
	if !ok {
		return nil, errf(CodeInclude, "unsupported file extension: %q", path)
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err // classified by the caller
	}
	if err := checkUTF8(path, data); err != nil {
		return nil, err
	}
	return parse(path, data)
}

// checkUTF8 enforces SPEC §2: file content anywhere that is not valid UTF-8 is
// E_PARSE. Stock parsers differ on whether they reject invalid bytes, replace
// them with U+FFFD, or pass them through, so the check is made up front.
func checkUTF8(path string, data []byte) error {
	if utf8.Valid(data) {
		return nil
	}
	return errf(CodeParse, "%s: content is not valid UTF-8", path)
}

// ---------------------------------------------------------------- JSON

// parseJSON walks the token stream instead of unmarshaling, because
// encoding/json silently keeps the last of a set of duplicate keys and SPEC §2
// requires E_PARSE. Numbers are decoded exactly (int64 when integral).
func parseJSON(path string, data []byte) (any, error) {
	dec := json.NewDecoder(bytes.NewReader(data))
	dec.UseNumber()
	v, err := jsonValue(dec, path)
	if err != nil {
		return nil, err
	}
	if _, err := dec.Token(); !errors.Is(err, io.EOF) {
		return nil, errf(CodeParse, "%s: trailing content after top-level JSON value", path)
	}
	return v, nil
}

func jsonValue(dec *json.Decoder, path string) (any, error) {
	tok, err := dec.Token()
	if err != nil {
		return nil, wrapf(CodeParse, err, "%s: invalid JSON", path)
	}
	return jsonValueFrom(dec, tok, path)
}

func jsonValueFrom(dec *json.Decoder, tok json.Token, path string) (any, error) {
	switch t := tok.(type) {
	case json.Delim:
		switch t {
		case '{':
			obj := make(map[string]any)
			for {
				keyTok, err := dec.Token()
				if err != nil {
					return nil, wrapf(CodeParse, err, "%s: invalid JSON", path)
				}
				if d, ok := keyTok.(json.Delim); ok && d == '}' {
					return obj, nil
				}
				key, ok := keyTok.(string)
				if !ok {
					return nil, errf(CodeParse, "%s: invalid JSON object key", path)
				}
				if _, dup := obj[key]; dup {
					return nil, errf(CodeParse, "%s: duplicate key %q in JSON object", path, key)
				}
				val, err := jsonValue(dec, path)
				if err != nil {
					return nil, err
				}
				obj[key] = val
			}
		case '[':
			arr := []any{}
			for {
				elemTok, err := dec.Token()
				if err != nil {
					return nil, wrapf(CodeParse, err, "%s: invalid JSON", path)
				}
				if d, ok := elemTok.(json.Delim); ok && d == ']' {
					return arr, nil
				}
				val, err := jsonValueFrom(dec, elemTok, path)
				if err != nil {
					return nil, err
				}
				arr = append(arr, val)
			}
		default:
			return nil, errf(CodeParse, "%s: unexpected %q in JSON", path, t)
		}
	case json.Number:
		n, err := numberFromString(t.String())
		if err != nil {
			return nil, errf(CodeParse, "%s: number %q has no JSON-equivalent value: %s", path, t.String(), err)
		}
		return n, nil
	case string, bool, nil:
		return t, nil
	default:
		return nil, errf(CodeParse, "%s: unexpected JSON token", path)
	}
}

// numberFromString keeps integers as int64 (no float64 rounding) and
// everything else as float64. A literal that does not land on a finite double
// has no value in the data model of SPEC §2 and is rejected: the tree stays
// JSON-encodable, and whole-value typing (SPEC §6) leaves such text a string.
func numberFromString(s string) (any, error) {
	if i, err := strconv.ParseInt(s, 10, 64); err == nil {
		return i, nil
	}
	f, err := strconv.ParseFloat(s, 64)
	if err != nil && !errors.Is(err, strconv.ErrRange) {
		return nil, err
	}
	if math.IsInf(f, 0) || math.IsNaN(f) {
		return nil, errors.New("out of IEEE-754 double range")
	}
	return f, nil
}

// ---------------------------------------------------------------- YAML

// parseYAML decodes the stream document by document: SPEC §2 allows at most
// one document per file, an empty document parses as null, and a
// multi-document stream is E_PARSE.
func parseYAML(path string, data []byte) (any, error) {
	dec := yaml.NewDecoder(bytes.NewReader(data))
	var doc *yaml.Node
	for {
		var node yaml.Node
		err := dec.Decode(&node)
		if errors.Is(err, io.EOF) {
			break
		}
		if err != nil {
			return nil, wrapf(CodeParse, err, "%s: invalid YAML", path)
		}
		if doc != nil {
			return nil, errf(CodeParse, "%s: multi-document YAML stream (at most one document per file)", path)
		}
		first := node
		doc = &first
	}
	if doc == nil || doc.Kind == 0 {
		return nil, nil // empty document (SPEC §2)
	}
	c := &yamlConv{path: path, active: map[*yaml.Node]bool{}}
	return c.value(doc)
}

// maxYAMLNodes is the alias-expansion budget of SPEC §2: a document whose
// fully expanded tree would exceed this many nodes is E_PARSE.
const maxYAMLNodes = 1_000_000

// yamlConv converts one YAML document.
//
// active holds the nodes on the current path, so an alias that reaches an
// ancestor is reported as the cycle SPEC §2 requires. nodes is the expansion
// budget: it counts the nodes actually produced — each scalar value, sequence
// element and mapping entry — as they are built, so an alias bomb (a layered
// anchor graph whose expansion is exponential in the source size) is rejected
// after a bounded amount of work rather than materialized. This is a real
// budget on output size, not a limit on depth or on the number of alias
// references: heavy but honest reuse expands to few nodes and loads.
type yamlConv struct {
	path   string
	active map[*yaml.Node]bool
	nodes  int
}

// charge accounts for n nodes about to be produced and fails once the
// document's expansion passes the budget.
func (c *yamlConv) charge(n int, at *yaml.Node) error {
	c.nodes += n
	if c.nodes > maxYAMLNodes {
		return errf(CodeParse,
			"%s:%d: expanded YAML tree exceeds the %d-node limit (alias expansion is bounded)",
			c.path, at.Line, maxYAMLNodes)
	}
	return nil
}

// yamlCoreTags is the tag set of the YAML 1.2 core schema. SPEC §2 makes any
// other tag (including 1.1 type-library tags such as !!timestamp) E_PARSE.
var yamlCoreTags = map[string]bool{
	"!!str": true, "!!int": true, "!!float": true, "!!bool": true,
	"!!null": true, "!!map": true, "!!seq": true,
}

// value converts one node, charging the expansion budget for every node it
// produces. An alias is expanded through Node.Alias, so the budget — not a
// depth heuristic — is what stops an exponential anchor graph.
func (c *yamlConv) value(n *yaml.Node) (any, error) {
	if c.active[n] {
		return nil, errf(CodeParse, "%s:%d: cyclic YAML alias", c.path, n.Line)
	}
	c.active[n] = true
	defer delete(c.active, n)

	switch n.Kind {
	case yaml.DocumentNode:
		if len(n.Content) == 0 {
			return nil, nil
		}
		return c.value(n.Content[0])
	case yaml.AliasNode:
		// SPEC §2: anchors and aliases are resolved at parse time and produce
		// plain values. The alias itself is not a node of the expanded tree —
		// what it expands to is, and that is charged below.
		if n.Alias == nil {
			return nil, errf(CodeParse, "%s: unresolved alias %q", c.path, n.Value)
		}
		return c.value(n.Alias)
	case yaml.MappingNode:
		if err := checkYAMLTag(n, "!!map", c.path); err != nil {
			return nil, err
		}
		obj := make(map[string]any, len(n.Content)/2)
		for i := 0; i+1 < len(n.Content); i += 2 {
			keyNode, valNode := n.Content[i], n.Content[i+1]
			if err := c.charge(1, keyNode); err != nil { // one mapping entry
				return nil, err
			}
			key, err := c.value(keyNode)
			if err != nil {
				return nil, err
			}
			ks, ok := key.(string)
			if !ok {
				// The tree is JSON-equivalent (SPEC §2): object keys are strings.
				return nil, errf(CodeParse, "%s:%d: non-string mapping key", c.path, keyNode.Line)
			}
			// SPEC §2: mapping keys are not counted — refund the scalar
			// charge value() just took for the key.
			c.nodes--
			if _, dup := obj[ks]; dup {
				return nil, errf(CodeParse, "%s:%d: duplicate key %q in mapping", c.path, keyNode.Line, ks)
			}
			v, err := c.value(valNode)
			if err != nil {
				return nil, err
			}
			obj[ks] = v
		}
		return obj, nil
	case yaml.SequenceNode:
		if err := checkYAMLTag(n, "!!seq", c.path); err != nil {
			return nil, err
		}
		if err := c.charge(len(n.Content), n); err != nil { // one per element
			return nil, err
		}
		arr := make([]any, 0, len(n.Content))
		for _, elem := range n.Content {
			v, err := c.value(elem)
			if err != nil {
				return nil, err
			}
			arr = append(arr, v)
		}
		return arr, nil
	case yaml.ScalarNode:
		if err := c.charge(1, n); err != nil { // one scalar value
			return nil, err
		}
		return yamlScalar(n, c.path)
	default:
		return nil, errf(CodeParse, "%s:%d: unsupported YAML node", c.path, n.Line)
	}
}

func checkYAMLTag(n *yaml.Node, want, path string) error {
	if n.Style&yaml.TaggedStyle == 0 {
		return nil // implicit tag, nothing was written in the document
	}
	if !yamlCoreTags[n.Tag] {
		return errf(CodeParse, "%s:%d: unsupported YAML tag %q", path, n.Line, n.Tag)
	}
	if n.Tag != want {
		return errf(CodeParse, "%s:%d: YAML tag %q does not apply to this node", path, n.Line, n.Tag)
	}
	return nil
}

// yamlScalar resolves a scalar under the YAML 1.2 core schema.
//
// yaml.v3 keeps several YAML 1.1 resolution rules (0777 as octal, 1_000 with
// underscore separators), so its own tag is deliberately ignored for plain
// scalars: the value text is re-resolved here with the core schema regexps.
// An explicit tag in the document is detectable via yaml.TaggedStyle and is
// honoured.
func yamlScalar(n *yaml.Node, path string) (any, error) {
	if n.Style&yaml.TaggedStyle != 0 {
		if !yamlCoreTags[n.Tag] {
			return nil, errf(CodeParse, "%s:%d: unsupported YAML tag %q", path, n.Line, n.Tag)
		}
		return yamlTagged(n, path)
	}
	// Quoted, literal and folded scalars are always strings.
	const quoted = yaml.DoubleQuotedStyle | yaml.SingleQuotedStyle | yaml.LiteralStyle | yaml.FoldedStyle
	if n.Style&quoted != 0 {
		return n.Value, nil
	}
	return checkFinite(yamlCoreResolve(n.Value), n, path)
}

func yamlTagged(n *yaml.Node, path string) (any, error) {
	switch n.Tag {
	case "!!str":
		return n.Value, nil
	case "!!null":
		return nil, nil
	case "!!bool":
		switch n.Value {
		case "true", "True", "TRUE":
			return true, nil
		case "false", "False", "FALSE":
			return false, nil
		}
		return nil, errf(CodeParse, "%s:%d: %q is not a core-schema boolean", path, n.Line, n.Value)
	case "!!int":
		if v, ok := yamlCoreResolve(n.Value).(int64); ok {
			return v, nil
		}
		return nil, errf(CodeParse, "%s:%d: %q is not a core-schema integer", path, n.Line, n.Value)
	case "!!float":
		switch v := yamlCoreResolve(n.Value).(type) {
		case float64:
			return checkFinite(v, n, path)
		case int64:
			return float64(v), nil
		}
		return nil, errf(CodeParse, "%s:%d: %q is not a core-schema float", path, n.Line, n.Value)
	}
	return nil, errf(CodeParse, "%s:%d: YAML tag %q does not apply to a scalar", path, n.Line, n.Tag)
}

// checkFinite rejects the core-schema floats that have no JSON-equivalent
// form: `.inf`, `.nan`, and any literal that overflows a double (SPEC §2).
func checkFinite(v any, n *yaml.Node, path string) (any, error) {
	if f, ok := v.(float64); ok && (math.IsInf(f, 0) || math.IsNaN(f)) {
		return nil, errf(CodeParse, "%s:%d: %q has no JSON-equivalent form", path, n.Line, n.Value)
	}
	return v, nil
}

// The YAML 1.2 core schema resolution regexps (spec §10.2.1.2).
var (
	yamlIntRe   = regexp.MustCompile(`^[-+]?[0-9]+$`)
	yamlOctRe   = regexp.MustCompile(`^0o[0-7]+$`)
	yamlHexRe   = regexp.MustCompile(`^0x[0-9a-fA-F]+$`)
	yamlFloatRe = regexp.MustCompile(`^[-+]?(\.[0-9]+|[0-9]+(\.[0-9]*)?)([eE][-+]?[0-9]+)?$`)
)

func yamlCoreResolve(v string) any {
	switch v {
	case "", "~", "null", "Null", "NULL":
		return nil
	case "true", "True", "TRUE":
		return true
	case "false", "False", "FALSE":
		return false
	case ".inf", ".Inf", ".INF", "+.inf", "+.Inf", "+.INF":
		return math.Inf(1)
	case "-.inf", "-.Inf", "-.INF":
		return math.Inf(-1)
	case ".nan", ".NaN", ".NAN":
		return math.NaN()
	}
	switch {
	case yamlIntRe.MatchString(v):
		if i, err := strconv.ParseInt(v, 10, 64); err == nil {
			return i
		}
		if f, ok := parseFloatOrInf(v); ok { // out of int64 range
			return f
		}
	case yamlOctRe.MatchString(v):
		if i, err := strconv.ParseInt(v[2:], 8, 64); err == nil {
			return i
		}
	case yamlHexRe.MatchString(v):
		if i, err := strconv.ParseInt(v[2:], 16, 64); err == nil {
			return i
		}
	case yamlFloatRe.MatchString(v):
		if f, ok := parseFloatOrInf(v); ok {
			return f
		}
	}
	return v
}

// parseFloatOrInf keeps an overflowing literal a *number* (±Inf) rather than
// silently demoting it to a string: the core schema resolved it as a number,
// and checkFinite then reports it as E_PARSE.
func parseFloatOrInf(v string) (float64, bool) {
	f, err := strconv.ParseFloat(v, 64)
	if err == nil || errors.Is(err, strconv.ErrRange) {
		return f, true
	}
	return 0, false
}

// ---------------------------------------------------------------- TOML

func parseTOML(path string, data []byte) (any, error) {
	var v any
	// go-toml/v2 reports duplicate keys and redefined tables as errors, which
	// is what SPEC §2 requires.
	if err := toml.Unmarshal(data, &v); err != nil {
		return nil, wrapf(CodeParse, err, "%s: invalid TOML", path)
	}
	return tomlNormalize(v, path)
}

// tomlNormalize converts the parts of go-toml's value model that are not
// JSON-equivalent: SPEC §2 requires datetimes to become RFC 3339-style
// strings, and TOML's `inf`/`nan` floats have no JSON form at all.
func tomlNormalize(v any, path string) (any, error) {
	switch t := v.(type) {
	case map[string]any:
		out := make(map[string]any, len(t))
		for k, val := range t {
			nv, err := tomlNormalize(val, path)
			if err != nil {
				return nil, err
			}
			out[k] = nv
		}
		return out, nil
	case []any:
		out := make([]any, len(t))
		for i, val := range t {
			nv, err := tomlNormalize(val, path)
			if err != nil {
				return nil, err
			}
			out[i] = nv
		}
		return out, nil
	case float64:
		if math.IsInf(t, 0) || math.IsNaN(t) {
			return nil, errf(CodeParse, "%s: %v has no JSON-equivalent form", path, t)
		}
		return t, nil
	case time.Time:
		// An offset date-time: uppercase T, a UTC offset written as "Z", any
		// other offset kept in its authored numeric form. RFC3339Nano also
		// drops trailing zeros from the fraction, and the "." with them.
		return t.Format(time.RFC3339Nano), nil
	case toml.LocalDateTime:
		return trimFraction(t.String()), nil
	case toml.LocalDate:
		return t.String(), nil
	case toml.LocalTime:
		return trimFraction(t.String()), nil
	}
	return v, nil
}

// trimFraction drops trailing zeros from a fractional-seconds group, and the
// "." along with them once the fraction reaches zero (SPEC §2). go-toml's
// local types render the fraction at the precision it was authored with.
func trimFraction(s string) string {
	dot := strings.IndexByte(s, '.')
	if dot < 0 {
		return s
	}
	end := dot + 1
	for end < len(s) && s[end] >= '0' && s[end] <= '9' {
		end++
	}
	frac := strings.TrimRight(s[dot+1:end], "0")
	if frac == "" {
		return s[:dot] + s[end:]
	}
	return s[:dot+1] + frac + s[end:]
}
